### PR TITLE
#185172113 user is able to see the recommended products to him/her

### PR DESCRIPTION
**Pull request by Kress2000**
-use search, buy history algorithms, and the newest added products to display the products to the buyer.
- a buyer will access the page, freely without logging in. [Delivers 185172113]

#### What does this PR do?
Fetches  products and displays them on home pages, and shows recommended products to the user so that he may be persued to  add them to the cart

#### Description of Task to be completed?

- The recommendation algorithm should take into account the buyer's purchase history and browsing behavior.
- Recommended products should be displayed with images and short product descriptions.
- Recommendations can help buyers discover new products and make informed purchasing decisions."

#### How should this be manually tested?
At home path  `/` you are able to view to the home page and navigate different products and news about the app

#### What are the relevant pivotal tracker/Trello stories?
#185172113

#### Screenshots (if appropriate):

https://github.com/atlp-rwanda/Team-Sostene-E-commerce-fe/assets/91186046/6c2861f6-4f9c-437f-b78e-ef90dec52071

### CODE REVIEW

File: vite.config.ts.timestamp-1686262896420-8b0154e28c0f1.mjs
Comment (AdrineUWERA) [inline vite.config.ts.timestamp-1686262896420-8b0154e28c0f1.mjs:1]: Remove this file

File: vite.config.ts
Comment (AdrineUWERA) [inline vite.config.ts:1]: slices and hooks need to be tested.

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]
The slider looks amazing.
image
The contrast here (card component) is kinda unbalanced.
I would suggest to change the button text to something like "shop". Something directly related to products.

File: vite.config.ts
Comment (Kress2000) [reply vite.config.ts:1]: done

File: src/pages/home/components/introCardproduct/introCard.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  category: string;
+  description: string;
+}
+
+export default function IntroCrads() {
+  const navigate = useNavigate();
+  const token = useSelector((state: RootState) => state.token.value);
+  const handleRedirectUser = () => (token ? navigate(routes.addToCart) : navigate(routes.login));
+  const { isLoading, data, error } = useFetch('https://fakestoreapi.com/products');
+  const newData: SingleProduct[] | null = data as SingleProduct[] | null;
```
Comment (Edcreation) [inline src/pages/home/components/introCardproduct/introCard.tsx:27]: You can use Data from backend, not dummy data

**Review by Edcreation — CHANGES_REQUESTED** [review]

File: src/pages/home/components/introCardproduct/introCard.tsx (hunk 1 above)
Comment (Edcreation) [reply src/pages/home/components/introCardproduct/introCard.tsx:27]: Also the Recommended product cards needs to be changed. you can refer to Figma design.
Screenshot 2023-06-10 180022

File: src/pages/home/components/slider/Slider.tsx
Comment (AdrineUWERA) [inline src/pages/home/components/slider/Slider.tsx:1]: This Slideshow has to fit the screen as it is on develop branch

File: src/pages/home/components/productMountainSlider/productMountainSlider.tsx
Comment (AdrineUWERA) [inline src/pages/home/components/productMountainSlider/productMountainSlider.tsx:1]: You can remove the white space on the sides

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]

File: src/pages/home/components/slider/Slider.tsx
Comment (Kress2000) [reply src/pages/home/components/slider/Slider.tsx:1]: no, i have to follow the Figma design, and obey the paddings

File: styles/mixins.scss [hunk 2]
```diff
@@ -74,7 +74,7 @@
 
   /* Handle */
```
Comment (Pacific911) [inline styles/mixins.scss:75]: same here

**Review by Pacific911 — CHANGES_REQUESTED** [review]

File: src/pages/home/components/introCardproduct/introCard.module.scss
Comment (AdrineUWERA) [inline src/pages/home/components/introCardproduct/introCard.module.scss:1]: image

make sure the space on the sides are equal

**Review by AdrineUWERA — APPROVED** [review]

**Review by Edcreation — CHANGES_REQUESTED** [review]
This card could be worked on more. why this black background?
Screenshot 2023-06-13 223521

File: src/pages/home/components/productMountainSlider/productMountainSlider.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  createdAt?: string;
+  expDate: string;
+  expiredflag: boolean;
+  name: string;
+  quantity: number;
+}
+
+interface ProductsProps {
+  imageData: newProduct[];
+}
```
Comment (princoo) [inline src/pages/home/components/productMountainSlider/productMountainSlider.tsx:26]: does `imageData` have this `newProduct` instances? (category,rating,.....)
may be you can change the name of `imageData` to give it an understandable meaning

File: src/pages/home/components/recommendedProducts/recommendedProducts.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    expiredflag: boolean;
+    name: string;
+    quantity: number;
+  };
+}
+export default function RecommendedProducts(props: ProductsProps) {
+  const { data } = props;
+  const navigate = useNavigate();
+  const token = useSelector((state: RootState) => state.token.value);
+  const handleRedirectUser = () => (token ? navigate(routes.addToCart) : navigate(routes.login));
```
Comment (princoo) [inline src/pages/home/components/recommendedProducts/recommendedProducts.tsx:28]: no need to use `useSelector `and define the type for it,
you may use the predefined hook `useAppSellector `  which does not require all that

**Review by princoo — CHANGES_REQUESTED** [review]

**Review by Pacific911 — CHANGES_REQUESTED** [review]
am not seeing which page it redirects you after you click on shop ?? look on it

File: src/pages/home/components/productMountainSlider/productMountainSlider.tsx (hunk 3 above)
Comment (Kress2000) [reply src/pages/home/components/productMountainSlider/productMountainSlider.tsx:26]: it is fine, they are the same

File: src/pages/accounts/signup/components/button/__test__/button.test.tsx [hunk 5]
```diff
@@ -16,4 +16,14 @@ describe('Button Element', () => {
...
+  it('renders the button component with default', () => {
+    render(<Button size="" color="" value="Click me" onclick={vi.fn()} type="button" />);
+
+    const buttonElement = screen.getByText('Click me') as HTMLButtonElement;
+    expect(buttonElement).toBeInTheDocument();
+    expect(screen.getByText('Click me')).toBeInTheDocument();
+    expect(buttonElement.tagName).toBe('BUTTON');
+    expect(buttonElement).toHaveAttribute('type', 'button');
+    fireEvent.click(buttonElement);
+  });
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply src/pages/accounts/signup/components/button/__test__/button.test.tsx:28]: wanted to cover the branch functions of launched that btn bcz they are many, and covered now

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]
The cards are not integrated. Make sure a product can be added to cart and also leads to the to product page when clicked.

**Review by Edcreation — COMMENTED** [review]
Those recommended product cards you can use the product card created in components. since it can fetch real product reviews.

**Review by Edcreation — APPROVED** [review]

**Review by JosephNgabo — APPROVED** [review]

**Comment by princoo** [comment]
adjust you styling on you header of `Recommended Products`
bug

**Review by princoo — COMMENTED** [review]
adjust your heading `Recommended Products `

bug

File: src/components/Product/TopProducts.tsx [hunk 6]
```diff
@@ -4,12 +4,13 @@ export default function TopProducts() {
   const products = useTopProducts();
 
   return (
-    <div className="pt-5">
-      {products.loading ? (
+    <div className="w-full">
+      {products.loading && (
         <div className="w-full p-2 text-center">
           <i className="fa fa-spinner fa-spin text-orange" aria-hidden="true"></i>
         </div>
```
Comment (princoo) [inline src/components/Product/TopProducts.tsx:11]: to maintain consistency in UI plz use the defined components for the spinner

**Review by princoo — APPROVED** [review]

File: src/pages/home/home.tsx [hunk 7]
```diff
@@ -1,15 +1,75 @@
...
+  page: number;
+  totalPages: number;
+}
 
 function Home() {
+  const url = `${BACKEND_URL}/products/all?page=1&limit=10`;
+  const { data } = useFetch(url);
+  const instantData: cardDetails | null = data as cardDetails | null;
+  const products = instantData?.products;
+
```
Comment (AdrineUWERA) [inline src/pages/home/home.tsx:36]: Can you create a separate component and not do the fetching and mapping inside this page file?

File: src/pages/home/home.tsx [hunk 8]
```diff
@@ -1,15 +1,75 @@
...
+      <div className="wrapperHome">
+        <div className="homeSlides" data-testid="home">
+          <HomeParallax />
+        </div>
+        <div className="newProducts">
+          <div className="title">New Release</div>
+          <div className="slides_of_Products">
+            {products && <ProductMountainSlider imageData={products} />}
+            <div className="over"></div>
+            <img src="./images/gif.gif" className="bg"></img>
```
Comment (AdrineUWERA) [inline src/pages/home/home.tsx:48]: I honestly think this gif background doesn't look good. I fetched this branch and removed the background and it looks better. Consider removing it.

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]
I don't know why u added whitespace around all page components. It doesn't look good. Have the white space around a specific component instead of the whole page. 

One more thing, arrange the home components in a way that makes sense. There is a lot going on on the page so having a logical order would help with the user experience. 

I would propose this order: 
1. HomePallax
2. HomeCards(They have a call to action as well as categories)
3. Recommended products
4. Top products
5. New release

**Comment by Edcreation** [comment]
> I don't know why u added whitespace around all page components. It doesn't look good. Have the white space around a specific component instead of the whole page.
> 
> image
> 
> One more thing, arrange the home components in a way that makes sense. There is a lot going on on the page so having a logical order would help with the user experience.
> 
> I would propose this order:
> 
> 1. HomePallax
> 2. HomeCards(They have a call to action as well as categories)
> 3. Recommended products
> 4. Top products
> 5. New release

Also, centering recommended products cards

**Review by Pacific911 — CHANGES_REQUESTED** [review]
I think you can reduce the white background on the image background to look better instead of placing it in the center

**Review by AdrineUWERA — CHANGES_REQUESTED** [review]
Add space between the components

**Review by munezerobagira — CHANGES_REQUESTED** [review]
the slide is  amazing you did great job, but I am pressing on the image on slide and the preview image is behind slider, though I suggest this, instead of preview we can make sure that   if  person click on image on active slide the he is redirected to view single product page.

**Review by AdrineUWERA — APPROVED** [review]

**Review by munezerobagira — APPROVED** [review]

**Pull request merged**